WalletWasabi/Transactions/TransactionStore.cs

```diff
@@ -0,0 +1,350 @@
+			}
+
+			return removed;
+		}
+
+		public bool TryGetTransaction(uint256 hash, out SmartTransaction sameStx)
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.TryGetValue(hash, out sameStx);
+			}
+		}
+
+		public ISet<SmartTransaction> GetTransactions()
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.Values.ToHashSet();
+			}
+		}
+
+		public ISet<uint256> GetTransactionHashes()
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.Keys.ToHashSet();
+			}
+		}
+
+		public bool IsEmpty()
+		{
+			lock (TransactionsLock)
+			{
+				return !Transactions.Any();
+			}
+		}
+
+		public bool Contains(uint256 hash)
+		{
+			lock (TransactionsLock)
+			{
+				return Transactions.ContainsKey(hash);
+			}
+		}
+
+		#region Serialization
+
+		private async Task SerializeAllTransactionsNoMutexAsync()
+		{
+			List<SmartTransaction> transactionsClone;
+			lock (TransactionsLock)
+			{
+				transactionsClone = Transactions.Values.ToList();
+			}
+
+			await TransactionsFileManager.WriteAllLinesAsync(transactionsClone.ToBlockchainOrderedLines()).ConfigureAwait(false);
+		}
+
+		private async Task TryAppendToFileAsync(params SmartTransaction[] transactions)
+			=> await TryAppendToFileAsync(transactions as IEnumerable<SmartTransaction>).ConfigureAwait(false);
+
+		private async Task TryAppendToFileAsync(IEnumerable<SmartTransaction> transactions)
+			=> await TryCommitToFileAsync(new Append(transactions)).ConfigureAwait(false);
+
+		private async Task TryRemoveFromFileAsync(params uint256[] transactionIds)
+			=> await TryRemoveFromFileAsync(transactionIds as IEnumerable<uint256>).ConfigureAwait(false);
+
+		private async Task TryRemoveFromFileAsync(IEnumerable<uint256> transactionIds)
+			=> await TryCommitToFileAsync(new Remove(transactionIds)).ConfigureAwait(false);
+
+		private List<ITxStoreOperation> Operations { get; } = new List<ITxStoreOperation>();
+		private object OperationsLock { get; } = new object();
+
+		private async Task TryCommitToFileAsync(ITxStoreOperation operation)
+		{
+			try
+			{
+				if (operation is null || operation.IsEmpty)
+				{
+					return;
+				}
+
+				// Make sure that only one call can continue.
+				lock (OperationsLock)
+				{
+					var isRunning = Operations.Any();
+					Operations.Add(operation);
+					if (isRunning)
+					{
+						return;
+					}
+				}
+
+				// Wait until the operation list calms down.
+				List<ITxStoreOperation> operationsToExecute;
+				while (true)
+				{
+					var count = Operations.Count;
+
+					await Task.Delay(100).ConfigureAwait(false);
+
+					lock (OperationsLock)
+					{
+						if (count == Operations.Count)
+						{
+							// Merge operations.
+							operationsToExecute = OperationMerger.Merge(Operations).ToList();
+							Operations.Clear();
+							break;
+						}
+					}
+				}
+
+				using (await TransactionsFileManager.Mutex.LockAsync().ConfigureAwait(false))
+				{
+					foreach (ITxStoreOperation op in operationsToExecute)
```

Comment:
I thought that the Merge will optimize the Additions according to Removes and vice versa, then here you check if there is at least one remove then set up the list and write the whole file in one iteration.  Otherwise if only addition then just append.

Reply:
That would make remove, remove from memory and nwould not remove from file.

Reply:
Ack.